### PR TITLE
SYCL: Add clang-tidy check

### PR DESCRIPTION
**Pull request by masterleinad**
Follow-up to #7530.

### CODE REVIEW

File: core/unit_test/sycl/TestSYCL_InterOp_Graph.cpp [hunk 1]
```diff
@@ -14,16 +14,20 @@
...
 
 #include <gtest/gtest.h>
 
+#include <fstream>
+#include <regex>
+
+#if !(defined(_GLIBCXX_RELEASE) && _GLIBCXX_RELEASE < 9) && \
+    !(defined(_LIBCPP_VERSION) && _LIBCPP_VERSION < 110000)
+#include <filesystem>
+#endif
```
Comment (masterleinad) [inline core/unit_test/sycl/TestSYCL_InterOp_Graph.cpp:29]: Trying to include `filesystem` caused problems with the underlying `clang` version used for `clang-tidy`.

Comment (dalg24) [reply core/unit_test/sycl/TestSYCL_InterOp_Graph.cpp:29]: Ok

**Review by dalg24 — APPROVED** [review]
I do not love the change to the Cuda half type header but I don't think it is worth re-running the tests

File: core/src/Cuda/Kokkos_Cuda_Half_Conversion.hpp
[the remark this answers is not in the document]
Comment (masterleinad) [reply core/src/Cuda/Kokkos_Cuda_Half_Conversion.hpp:1]: I needed that to make the CI happy.

[the remark this answers is not in the document]
Comment (dalg24) [reply core/src/Cuda/Kokkos_Cuda_Half_Conversion.hpp:1]: How so?  Was it something about the signature of the definition not matching the declaration?

File: core/src/impl/KokkosExp_IterateTileGPU.hpp [hunk 2]
```diff
@@ -875,7 +875,7 @@ struct is_array_type<T[]> : std::true_type {
 template <typename T>
 using value_type_storage_t =
     typename std::conditional_t<is_array_type<T>::value, std::decay<T>,
-                                std::add_lvalue_reference<T> >::type;
+                                std::add_lvalue_reference<T>>::type;
```
Comment (tpadioleau) [inline core/src/impl/KokkosExp_IterateTileGPU.hpp:878]: What do you think of the following ?

```cpp
template <typename T>
using value_type_storage_t =
    std::conditional_t<is_array_type<T>::value, std::decay_t<T>,
                                std::add_lvalue_reference_t<T>>;
```

File: core/unit_test/sycl/TestSYCL_InterOp_Graph.cpp (hunk 1 above)
Comment (tpadioleau) [reply core/unit_test/sycl/TestSYCL_InterOp_Graph.cpp:29]: I am surprised we are using an old clang here, doesn't it come from the oneapi install ? I would expect a recent clang/clang-tidy ?

File: core/src/impl/Kokkos_Half_FloatingPointWrapper.hpp
Comment (tpadioleau) [inline core/src/impl/Kokkos_Half_FloatingPointWrapper.hpp:1]: How does half type works ? This code was disabled in the other builds ?

**Review by tpadioleau — COMMENTED** [review]
Looks better :)

File: core/unit_test/sycl/TestSYCL_InterOp_Graph.cpp (hunk 1 above)
Comment (masterleinad) [reply core/unit_test/sycl/TestSYCL_InterOp_Graph.cpp:29]: I could still use an older stdlibc++. I haven't investigated further, though.

File: core/src/impl/Kokkos_Half_FloatingPointWrapper.hpp
Comment (masterleinad) [reply core/src/impl/Kokkos_Half_FloatingPointWrapper.hpp:1]: For most configurations the half types are just aliases to `float`. Only `Cuda`, `HIP` and `SYCL` use special types.

File: core/src/impl/KokkosExp_IterateTileGPU.hpp (hunk 2 above)
Comment (masterleinad) [reply core/src/impl/KokkosExp_IterateTileGPU.hpp]: Sure, see https://github.com/kokkos/kokkos/pull/7539/commits/2980f51adb93388e4feb02567273b884b42ce320.

**Review by Rombur — APPROVED** [review]

**Review by tpadioleau — APPROVED** [review]

File: core/unit_test/sycl/TestSYCL_InterOp_Graph.cpp (hunk 1 above)
Comment (tpadioleau) [reply core/unit_test/sycl/TestSYCL_InterOp_Graph.cpp:29]: I didn't see that the file had already such `ifdef` logic in the tests

**Comment by masterleinad** [comment]
I had to fix `HIP` in the same way as `Cuda`.

**Comment by dalg24** [comment]
Retest this please

**Pull request merged**